### PR TITLE
Feat/projector no pg transaction

### PR DESCRIPTION
**Pull request by AngelCastilloB**
# Context

Initial projection implementation stored all blocks in stability window buffer, but it has been optimized to only store it once close to tip since then. We can optimize it further to not even write to Block table (headers). This is a low-hanging fruit with high potential for performance gains.

# Proposed Solution

Implement a set of predicates that determines where a specific store will apply changes to the database or not.

### CODE REVIEW

File: packages/cardano-services/src/Projection/prepareTypeormProjection.ts [hunk 1]
```diff
@@ -347,7 +382,9 @@ export const prepareTypeormProjection = (
     entities: selectedEntities,
     extensions,
     mappers: selectedMappers,
-    stores: selectedStores
+    stores: selectedStores,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    willStore: <T>(evt: T) => willStoreCheckers.some((check) => check(evt as any))
```
Comment (mkazlauskas) [inline packages/cardano-services/src/Projection/prepareTypeormProjection.ts:387]: Please add some test(s) for this, doesn't have to be exhaustive

File: packages/cardano-services/src/Projection/createTypeormProjection.ts [hunk 2]
```diff
@@ -118,15 +118,23 @@ export const createTypeormProjection = ({
...
-        shareRetryBackoff(
-          (evt$) =>
-            evt$.pipe(
-              withTypeormTransaction({ connection$: connect() }),
-              applyStores(stores),
-              buffer.storeBlockData(),
-              typeormTransactionCommit()
-            ),
-          { shouldRetry: isRecoverableTypeormError }
+        // if there are any relevant data to write into db
```
Comment (mkazlauskas) [inline packages/cardano-services/src/Projection/createTypeormProjection.ts:121]: Please add a test/assertion that find an empty block in the dataset and asserts that blocks table doesn't contain it

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/cardano-services/src/Projection/prepareTypeormProjection.ts (hunk 1 above)
Comment (mkazlauskas) [inline packages/cardano-services/src/Projection/prepareTypeormProjection.ts:387]: Some typeorm store operators rely on postgresql 'on delete cascade' to perform the rollback changes (delete stuff), therefore we need to process such events
```suggestion
    willStore: <T extends ProjectionEvent>(evt: T) => evt.type === ChainSyncEventType.RollBackward || willStoreCheckers.some((check) => check(evt as any))
```

File: packages/projection-typeorm/src/operators/storeHandles.ts [hunk 3]
```diff
@@ -230,6 +230,8 @@ const withTotalSupplies = (
     )
   );
 
+export const willStoreHandles = ({ handles }: Mappers.WithHandles) => handles.length > 0;
```
Comment (mkazlauskas) [inline packages/projection-typeorm/src/operators/storeHandles.ts:233]: This seems to be missing test(s)

File: packages/projection-typeorm/test/operators/storeNftMetadata.test.ts [hunk 4]
```diff
@@ -447,3 +449,41 @@ describe('storeNftMetadata', () => {
...
+        cip67: {
+          byAssetId: {},
+          byLabel: {}
+        },
+        nftMetadata: [{} as ProjectedNftMetadata]
+      })
+    ).toBeTruthy();
+  });
+
+  it('returns true if there are cip67 metadata', () => {
```
Comment (mkazlauskas) [inline packages/projection-typeorm/test/operators/storeNftMetadata.test.ts:466]: These are the tokens without metadata that can potentially have reference NFTs minted in a previous block
```suggestion
  it('returns true if there are cip67 tokens', () => {
```

File: packages/projection-typeorm/src/operators/storePoolMetricsUpdateJob.ts [hunk 5]
```diff
@@ -3,6 +3,9 @@ import { STAKE_POOL_METRICS_UPDATE } from '../pgBoss';
 import { WithPgBoss } from './withTypeormTransaction';
 import { typeormOperator } from './util';
 
+export const willStorePoolMetricsUpdate = ({ eventType }: { eventType: ChainSyncEventType }) =>
+  eventType !== ChainSyncEventType.RollBackward;
```
Comment (mkazlauskas) [inline packages/projection-typeorm/src/operators/storePoolMetricsUpdateJob.ts:7]: We don't have to run this for all blocks, we can either
- use the same logic as in the store operator: once reached the tip and every `jobFrequency` blocks
- always return `false` and update the store operator logic to insert the job on 1st suitable block we see, e.g. if last seen block is 999, jobFrequency=1000, and we see block number 1111 then we would insert the job.

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/cardano-services/src/Projection/prepareTypeormProjection.ts (hunk 1 above)
Comment (AngelCastilloB) [reply packages/cardano-services/src/Projection/prepareTypeormProjection.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/b16dab93e2ae49680cd1d6ef47b76b5823ce0cd0

File: packages/projection-typeorm/src/operators/storeHandles.ts (hunk 3 above)
Comment (AngelCastilloB) [reply packages/projection-typeorm/src/operators/storeHandles.ts:233]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/b16dab93e2ae49680cd1d6ef47b76b5823ce0cd0

File: packages/projection-typeorm/test/operators/storeNftMetadata.test.ts (hunk 4 above)
Comment (AngelCastilloB) [reply packages/projection-typeorm/test/operators/storeNftMetadata.test.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/54c0c852e728eee3480f1244000a8165135b5455

File: packages/projection-typeorm/src/operators/storePoolMetricsUpdateJob.ts (hunk 5 above)
Comment (AngelCastilloB) [reply packages/projection-typeorm/src/operators/storePoolMetricsUpdateJob.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/36f6b38e783c9ac91eea234884c3153013f6f018

Comment (mkazlauskas) [reply packages/projection-typeorm/src/operators/storePoolMetricsUpdateJob.ts]: Please amend the commit message to mention this change of `storeStakePoolMetricsUpdateJob` behavior

File: packages/cardano-services/src/Projection/prepareTypeormProjection.ts (hunk 1 above)
Comment (AngelCastilloB) [reply packages/cardano-services/src/Projection/prepareTypeormProjection.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/f0d303377b97a236077dec1df3dce5eebe228aef , let me know if this is sufficient

Comment (mkazlauskas) [reply packages/cardano-services/src/Projection/prepareTypeormProjection.ts]: Thanks, looks good

File: packages/cardano-services/src/Projection/createTypeormProjection.ts (hunk 2 above)
Comment (AngelCastilloB) [reply packages/cardano-services/src/Projection/createTypeormProjection.ts:121]: Added the test here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/4bfa2c0db8a8f7ac8d777c77478f0161f567c1a6 , but implemented it a bit different, I created a test file with a few blocks and only one with mint, and then asserted only one block was stored

Comment (AngelCastilloB) [reply packages/cardano-services/src/Projection/createTypeormProjection.ts:121]: updated here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/3c09a5575b5348845768ecb56bea8b495f3327a5

File: packages/projection-typeorm/src/operators/storePoolMetricsUpdateJob.ts (hunk 5 above)
Comment (AngelCastilloB) [reply packages/projection-typeorm/src/operators/storePoolMetricsUpdateJob.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/307254250a1c35ea1fcab16287218672aacab9e8

**Review by mkazlauskas — DISMISSED** [review]
Great work! :rocket:

File: packages/projection-typeorm/src/operators/storeStakePoolMetadataJob.ts [hunk 6]
```diff
@@ -4,6 +4,12 @@ import { STAKE_POOL_METADATA_QUEUE, StakePoolMetadataJob, defaultJobOptions } fr
 import { WithPgBoss } from './withTypeormTransaction';
 import { certificatePointerToId, typeormOperator } from './util';
 
+export const willStoreStakePoolMetadataJob = ({
+  eventType,
+  stakePools
+}: { eventType: ChainSyncEventType } & Mappers.WithStakePools) =>
+  eventType !== ChainSyncEventType.RollBackward && stakePools.updates.length > 0;
```
Comment (iccicci) [inline packages/projection-typeorm/src/operators/storeStakePoolMetadataJob.ts:11]: For consistency I would change this in `eventType === ChainSyncEventType.RollForward`

**Review by iccicci — CHANGES_REQUESTED** [review]
I can see some `willStore*` check is missing the `eventType === ChainSyncEventType.RollForward` check.

I marked only one of them, next I have the doubt: is this as intended and in affirmative case why?

File: packages/projection-typeorm/src/operators/storeAssets.ts [hunk 7]
```diff
@@ -53,6 +53,8 @@ const rollBackward = async ({ mint, queryRunner }: StoreAssetEventParams): Promi
   return mintedAssetTotalSupplies;
 };
 
+export const willStoreAssets = ({ mint }: Mappers.WithMint) => mint.length > 0;
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/projection-typeorm/src/operators/storeAssets.ts:56]: It's ok to process all RollBackward blocks (not many of them). For consistency and simplicity I would remove all checks instead

[the remark this answers is not in the document]
Comment (AngelCastilloB) [reply packages/projection-typeorm/src/operators/storeAssets.ts:56]: Some of the stores alter the database on roll backward and forward, I.E:

```
  async ({ handles, queryRunner, eventType, block, mintedAssetTotalSupplies, handleMetadata }) => {
    const handleEventParams: HandleEventParams = {
      block,
      handleMetadata,
      handles: await withTotalSupplies(queryRunner, handles, mintedAssetTotalSupplies),
      queryRunner
    };

    eventType === ChainSyncEventType.RollForward
      ? await rollForward(handleEventParams)
      : await rollBackward(handleEventParams);
  }
```
while some other stores only perform side effects on roll forward, so the predicates usually follow that (what ever the logic is inside this tore that leads to a side effect on the database)

File: packages/projection-typeorm/src/operators/storeStakePoolMetadataJob.ts (hunk 6 above)
Comment (AngelCastilloB) [reply packages/projection-typeorm/src/operators/storeStakePoolMetadataJob.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/2af6d006fcc947890e054bbb1fc4f437f54c88cb

File: packages/projection-typeorm/src/operators/storeAssets.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (AngelCastilloB) [reply packages/projection-typeorm/src/operators/storeAssets.ts:56]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1200/commits/2af6d006fcc947890e054bbb1fc4f437f54c88cb

**Review by iccicci — APPROVED** [review]
Great job

**Review by mkazlauskas — APPROVED** [review]
